File: jvm/core/main/scala/com/webank/eggroll/core/containers/meta/DownloadContainersRequest.scala
package com.webank.eggroll.core.containers.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.meta.Containers

import scala.collection.JavaConverters._
import scala.language.implicitConversions

case class DownloadContainersRequest(
                                      sessionId: String,
                                      ranks: Array[Int],
                                      compressMethod: String = "zip",
                                      compressLevel: Int = 1,
                                      contentType: ContentType.ContentType
                                    )


object DownloadContainersRequest {
  implicit def deserialize(byteString: ByteString): DownloadContainersRequest = {
    val src = Containers.DownloadContainersRequest.parseFrom(byteString)
    DownloadContainersRequest(
      sessionId = src.getSessionId,
      ranks = src.getRanksList.asScala.map(_.toInt).toArray,
      compressMethod = src.getCompressMethod,
      compressLevel = src.getCompressLevel,
      contentType = src.getContentType
    )
  }

  implicit def serialize(src: DownloadContainersRequest): Array[Byte] = {
    val builder = Containers.DownloadContainersRequest.newBuilder()
      .setSessionId(src.sessionId)
      .addAllRanks(src.ranks.map(_.asInstanceOf[java.lang.Integer]).toList.asJava)
      .setCompressMethod(src.compressMethod)
      .setCompressLevel(src.compressLevel)
      .setContentType(src.contentType)
    builder.build().toByteArray
  }
}

File: jvm/new_core/src/main/java/com/eggroll/core/pojo/DownloadContainersRequest.java
package com.eggroll.core.pojo;

import com.webank.eggroll.core.meta.Containers;
import lombok.Data;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

@Data
public class DownloadContainersRequest implements RpcMessage {

    Logger log = LoggerFactory.getLogger(ContainerContent.class);

    private String sessionId;
    private String compressMethod;
    private List<Integer> ranks;
    private int compressLevel;
    private Containers.ContentType contentType;

    public DownloadContainersRequest(String sessionId, String compressMethod, List<Integer> ranks, int compressLevel, Containers.ContentType contentType) {
        this.sessionId = sessionId;
        this.compressMethod = compressMethod;
        this.ranks = ranks;
        this.compressLevel = compressLevel;
        this.contentType = contentType;
    }

    public Containers.DownloadContainersRequest toProto() {
        Containers.DownloadContainersRequest.Builder builder = Containers.DownloadContainersRequest.newBuilder();
        if (this.sessionId != null) {
            builder.setSessionId(this.sessionId);
        }
        if (this.compressMethod != null) {
            builder.setCompressMethod(this.compressMethod);
        }
        builder.setCompressLevel(this.compressLevel);
        if (this.contentType != null) {
            builder.setContentType(this.contentType);
        }
        return builder.build();
    }

    @Override
    public byte[] serialize() {
        return toProto().toByteArray();
    }

    @Override
    public void deserialize(byte[] data) {
        try {
            Containers.DownloadContainersRequest src = Containers.DownloadContainersRequest.parseFrom(data);
            this.sessionId = src.getSessionId();
            this.compressMethod = src.getCompressMethod();
            this.compressLevel = src.getCompressLevel();
            this.contentType = src.getContentType();
        } catch (Exception e) {
            log.error("deserialize error : ", e);
        }
    }
}
